ibc/store/bulk_store_cache.go

```diff
@@ -0,0 +1,188 @@
+	return func(m modules.BulkStoreCacher) {
+		if mod, ok := m.(*bulkStoreCache); ok {
+			mod.logger = logger
+		}
+	}
+}
+
+// WithPrivateKey assigns the private key to the BulkStoreCacher
+func WithPrivateKey(privateKeyHex string) modules.BulkStoreCacherOption {
+	return func(m modules.BulkStoreCacher) {
+		if mod, ok := m.(*bulkStoreCache); ok {
+			mod.privateKey = privateKeyHex
+		}
+	}
+}
+
+// WithStoresDir assigns the stores directory to the BulkStoreCacher
+func WithStoresDir(storesDir string) modules.BulkStoreCacherOption {
+	return func(m modules.BulkStoreCacher) {
+		if mod, ok := m.(*bulkStoreCache); ok {
+			mod.storesDir = storesDir
+		}
+	}
+}
+
+func (*bulkStoreCache) Create(bus modules.Bus, config *configs.BulkStoreCacherConfig, options ...modules.BulkStoreCacherOption) (modules.BulkStoreCacher, error) {
+	s := &bulkStoreCache{
+		cfg: config,
+	}
+	for _, option := range options {
+		option(s)
+	}
+	s.logger.Info().Msg("💾 Creating Bulk Store Cacher 💾")
+	bus.RegisterModule(s)
+	s.m = sync.Mutex{}
+	s.stores = make(map[string]*provableStore)
+	return s, nil
+}
+
+func (s *bulkStoreCache) GetModuleName() string { return modules.BulkStoreCacherModuleName }
+
+// AddStore creates and adds a provableStore to the bulkStoreCache
+// if one of the same name does not already exist
+func (s *bulkStoreCache) AddStore(name string) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if _, ok := s.stores[name]; ok {
+		return coreTypes.ErrIBCStoreAlreadyExists(name)
+	}
+	store := newProvableStore(s.GetBus(), coreTypes.CommitmentPrefix(name), s.privateKey)
+	s.stores[store.name] = store
+	return nil
+}
+
+// GetStore returns the provableStore with the given name
+func (s *bulkStoreCache) GetStore(name string) (modules.ProvableStore, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	store, ok := s.stores[name]
+	if !ok {
+		return nil, coreTypes.ErrIBCStoreDoesNotExist(name)
+	}
+	return store, nil
+}
+
+// RemoveStore removes the provableStore with the given name
+func (s *bulkStoreCache) RemoveStore(name string) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if _, ok := s.stores[name]; !ok {
+		return coreTypes.ErrIBCStoreDoesNotExist(name)
+	}
+	delete(s.stores, name)
+	return nil
+}
+
+// GetAllStores returns the map of stores to their store names
+func (s *bulkStoreCache) GetAllStores() map[string]modules.ProvableStore {
+	s.m.Lock()
+	defer s.m.Unlock()
+	stores := make(map[string]modules.ProvableStore, len(s.stores))
+	for name, store := range s.stores {
+		stores[name] = store
+	}
+	return stores
+}
+
+// FlushAllEntries caches all the entries for all stores in the bulkStoreCache
+func (s *bulkStoreCache) FlushAllEntries() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.logger.Info().Msg("🚽 Flushing All Cache Entries to Disk 🚽")
+	disk, err := newKVStore(s.storesDir)
+	if err != nil {
+		return err
+	}
+	for _, store := range s.stores {
+		if err := store.FlushEntries(disk); err != nil {
```

Comment:
Consider adding a log line for each store flushed (with a reference/id as to which one it is) so we can track it easily in case a single one fails.

Potentially just on error.

Ditto below for pruning.